pkg/cloudprovider/providers/aws/volumecache.go

```diff
@@ -0,0 +1,265 @@
+type volumeCache struct {
+	ec2 EC2
+
+	mutex  sync.RWMutex
+	latest *volumeCacheSnapshot
+}
+
+// volumeCacheSnapshot holds a view at a moment in time of all the volumes
+type volumeCacheSnapshot struct {
+	timestamp  int64
+	byVolumeId map[awsVolumeID]*cachedVolume
+}
+
+// cachedInstance is a reusable volume wrapper, holding some immutable state and caching access to mutable state
+type cachedVolume struct {
+	ID awsVolumeID
+
+	// Values that cannot change are exported publicly
+	AvailabilityZone string
+
+	cache *volumeCache
+
+	// mutex protects the mutable elements below
+	mutex sync.Mutex
+
+	// state is the state in EC2 as observed last
+	state *ec2.Volume
+
+	// stateTimestamp is the timestamp at which we last updated state
+	stateTimestamp int64
+}
+
+// DescribeInstance gets the state for the volume, subject to the cache policy.
+// Currently a cache-miss will cause a refresh of all volumes, because this is the same "price" in terms of quota.
+func (i *cachedVolume) DescribeVolume(cachePolicy *CachePolicy) (*ec2.Volume, error) {
+	snapshot := i.cache.latestSnapshot()
+
+	if snapshot != nil {
+		instance := snapshot.findCachedVolume(i.ID)
+		if instance != nil {
+			state := i.findStateIfCached(cachePolicy)
+			if state != nil {
+				return state, nil
+			} else {
+				glog.V(2).Infof("triggering volume refresh because state expired (volume %s, %s)", i.ID, cachePolicy.Name)
+			}
+		}
+	}
+
+	updated, err := i.cache.refreshAll(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	instance := updated.findCachedVolume(i.ID)
+	if instance == nil {
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	state := i.findStateIfCached(nil)
+	if state != nil {
+		return state, nil
+	}
+
+	return nil, cloudprovider.InstanceNotFound
+}
+
+// findStateIfCached returns the last state, if it is valid under the cache policy.
+// If there is no state, or the state has expired, it returns nil.
+func (i *cachedVolume) findStateIfCached(cachePolicy *CachePolicy) *ec2.Volume {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.state != nil {
+		if cachePolicy == nil || cachePolicy.IsValid(i.stateTimestamp) {
+			return i.state
+		} else {
+			glog.V(2).Infof("triggering refresh because cache policy %s expired", cachePolicy.Name)
+		}
+	}
+
+	return nil
+}
+
+// newInstanceCache is a constructor for an volumeCache
+func newVolumeCache(ec2 EC2) *volumeCache {
+	return &volumeCache{
+		ec2: ec2,
+	}
+}
+
+// latestSnapshot returns the last view of all volumes (or nil if no snapshot)
+func (c *volumeCache) latestSnapshot() *volumeCacheSnapshot {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return c.latest
+}
+
+// refreshAll builds an updated snapshot by calling DescribeVolume.
+// It accepts the previous snapshot, and will not refresh if a concurrent request has triggered a concurrent refresh.
+func (c *volumeCache) refreshAll(previous *volumeCacheSnapshot) (*volumeCacheSnapshot, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
refreshAll might takes a little while to finish. During this time, cache will be locked totally, so other operations will be blocked? I think during ec2 describeInsances or describeVolumes, reading cache  should not be block?